### ISSUE TITLE
Nothing prevents concurrent writes (or concurrent reads)

### ISSUE THREAD

**Issue by itamarst** [mitsuhiko/tokio-unix-ipc#16]
I may be misunderstanding the code, but given the `send()` methods aren't taking `&mut self`, it seems like nothing prevents concurrent or interleaved writing, which can result in corrupted messages.

(I was seeing symptoms of this, wrapped Sender with tokio Mutex and I _think_ it fixed it. But, you know, race conditions, could be unrelated.)

**Comment by mitsuhiko** [comment]
That's correct. I'm not sure if this library should try to do something about this, or if that's up to the user.

**Comment by itamarst** [comment]
I can imagine people wanting their own locking patterns; maybe you want a lock per `Sender`, or maybe you are parent process writing to all client processes each time, one of my use cases, and so you want a single lock for multiple `Sender`s.

But the current interface is misleading, and can lead to silent data corruption that idiomatically would've been prevented via the type system. So I would suggest `Sender::sender()` and `recv()` be changed to accept `&mut self`, to match the standard Rust pattern (e.g. `Read` and `Write` traits).

**Comment by mitsuhiko** [comment]
I really don't like the idea of forcing the this to use `&mut self` as it makes the type much more frustrating to work with in some common patterns.

**Comment by mitsuhiko** [comment]
@cgwalters what do you think?

**Comment by cgwalters** [comment]
Well I got nerd sniped into this https://github.com/mitsuhiko/tokio-unix-ipc/pull/17

**Comment by bkoropoff** [comment]
I almost got bitten by this before thinking to double-check that sends are atomic.  IMO, `Sender` and `Receiver` should both take `&mut self` and have wrappers that take `&self` and are internally synchronized.
